Enable hipBLASLt

ahsan-ca opened the pull request:
This PR adds functionality to enable hipBLASLt in MIGraphX.
hipBLASLt can be enabled via MIGRAPHX_ENABLE_HIPBLASLT_GEMM env variable.

pfultz2 commented on src/targets/gpu/include/migraphx/gpu/context.hpp:
```diff
@@ -126,6 +127,41 @@ struct hip_device
+            }
+            assert(hbltworkspace.get() != nullptr);
+            return hbltworkspace.get();
+        }
```
The workspace should not be stored in the context. Instead during lowering we can add a allocate instruction and then pass it as an argument to the operator. This is the way we handle the workspace in miopen.

ahsan-ca replied:
Working on this...

ahsan-ca replied:
Made the changes to add `allocate_gpu` in lowering.

ahsan-ca replied:
So the change in lowering for `allocate_gpu` would not work. I have removed the change and I will discuss this with Paul when he is back to see what we should do here.

pfultz2 replied:
What the issue that you ran into?

pfultz2 replied:
The workspace should be added before the output parameter.

ahsan-ca replied:
Done.

pfultz2 commented on src/targets/gpu/include/migraphx/gpu/context.hpp:
```diff
@@ -126,6 +127,41 @@ struct hip_device
+                hbltpreference = create_hipblaslt_preference_ptr();
+            }
+            assert(hbltpreference.get() != nullptr);
+            return hbltpreference.get();
```
The preference pointer should be stored in the operator and created either during compile_blas or finalize.

pfultz2 replied:
Actually it should just be created in the solution class. anyways.

ahsan-ca replied:
Done.

pfultz2 commented on src/targets/gpu/hip_gemm_impl.cpp:
```diff
@@ -0,0 +1,599 @@
+        CHECK_HIPBLAS_ERROR(hipblasLtMatmulDescDestroy(hipblaslt_desc));
+        CHECK_HIPBLAS_ERROR(hipblasLtMatrixLayoutDestroy(matA));
+        CHECK_HIPBLAS_ERROR(hipblasLtMatrixLayoutDestroy(matB));
+        CHECK_HIPBLAS_ERROR(hipblasLtMatrixLayoutDestroy(matC));
```
Use managed pointers.

ahsan-ca replied:
I was running into issues (test case failures) with using managed pointers for this. Need to look deeper into this. Any recommendations(example) on how to use managed pointers for this are welcome :).

pfultz2 commented on src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:
```diff
@@ -0,0 +1,67 @@
+#include <migraphx/gpu/config.hpp>
+#if MIGRAPHX_USE_HIPBLASLT
+#include <hipblaslt/hipblaslt.h>
+#define HIPBLASLT_WORKSPACE_SIZE (2 * 128 * 1024 * 1024)
```
Create a constant variable, dont use a macro.

ahsan-ca replied:
Done.

pfultz2 commented on src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:
```diff
@@ -0,0 +1,67 @@
+#include <hipblaslt/hipblaslt.h>
+#define HIPBLASLT_WORKSPACE_SIZE (2 * 128 * 1024 * 1024)
+#ifndef CHECK_HIPBLAS_ERROR
+#define CHECK_HIPBLAS_ERROR(error)                    \
```
Create a hipblaslt_invoke function to call the API and check the error code.

ahsan-ca replied:
Done.

pfultz2 commented:
You need to enable hipblaslt on the tidy builds as well.

ahsan-ca commented on src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:
```diff
@@ -0,0 +1,67 @@
+                error,                                \
+                __FILE__,                             \
+                __LINE__);                            \
+        exit(EXIT_FAILURE);                           \
```
Done.

ahsan-ca commented on src/targets/gpu/compile_blas.cpp:
```diff
@@ -0,0 +1,80 @@
+        auto alloc =
+            m.insert_instruction(ins, make_op("allocate", {{"shape", to_value(ins->get_shape())}}));
+        inputs.insert(std::prev(inputs.end()), alloc);
+        m.replace_instruction(ins, op, inputs);
```
Removed `compile_blas`.

umangyadav commented on src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:
```diff
@@ -25,22 +25,46 @@
-                __FILE__,                             \
-                __LINE__);                            \
-        exit(EXIT_FAILURE);                           \
+#ifndef HIPBLASLT_INVOKE_FUNCTION
+#define HIPBLASLT_INVOKE_FUNCTION
```
remove these ifdefs

ahsan-ca replied:
Done. I have also addressed the other comments that we discussed, in person, on Monday. Can you please also take a look at the last few commits to see if the comments have been addressed correctly? Thanks a lot for the help!

causten commented:
testing with this branch is working.  Just need some approvals.  @TedThemistokleous @CharlieL7

TedThemistokleous commented:
@ahsan-ca make sure we don't also need hipblas, not just hipblaslt and -dev/devel equivalent like we had in the SLES docker case if we're still missing deps.

ahsan-ca commented:
> > testing with this branch is working. Just need some approvals. @TedThemistokleous @CharlieL7
> 
> @causten I discussed the changes (made for addressing comments) with Umang yesterday, I need to update some things in these changes and fix some CI failures too. I will update here once done. Thanks!

I have made all the required changes and updated the PR with latest develop. This is good to go after CI run is green. Thanks! @CharlieL7 its ready for review. Thanks!

pfultz2 commented on src/targets/gpu/lowering.cpp:
```diff
@@ -251,7 +252,12 @@ struct miopen_apply
             assert(refs.size() == 2);
             auto output = insert_allocation(ins, ins->get_shape());
```
You need to do another `insert_allocation` for the workspace as well.

ahsan-ca replied:
Made changes for allocating workspace in lowering.

ahsan-ca commented:
> You need to enable hipblaslt on the tidy builds as well.

@pfultz2 I was looking at this and the cmake argument for hipblaslt is turned on by default `option(MIGRAPHX_USE_HIPBLASLT "Enable MIGraphX to use hipBLASLt" ON)`. 
`MIGRAPHX_ENABLE_HIPBLASLT_GEMM` is an runtime environment variable that controls turning hipblaslt on/off for code.
In this case, we probably don't need to add anything to the `ci.yaml` file for CI.
Is my understanding correct?

pfultz2 commented on hip-clang.docker:
```diff
@@ -31,6 +31,7 @@ RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --allow-
     libnuma-dev \
     miopen-hip \
     rocblas \
+    hipblaslt \
```
This should be installed in the install_preqs.sh file, not in the dockerfile.

ahsan-ca replied:
Moved to install_prereqs file.

pfultz2 commented on src/targets/gpu/CMakeLists.txt:
```diff
@@ -46,6 +49,15 @@ else()
+if(MIGRAPHX_USE_HIPBLASLT)
+    # hipblaslt
+    find_package(hipblaslt REQUIRED)
+    find_package(hipblas REQUIRED)
```
Why are we looking for hipblas? We dont use it.

ahsan-ca replied:
Removed.

ahsan-ca replied:
Seems like hipblaslt uses hipblas. I saw sles failure after removing hipblas:
```
/opt/rocm-6.1.3/include/hipblaslt/hipblaslt.h:49:10: fatal error: 'hipblas/hipblas.h' file not found
   49 | #include <hipblas/hipblas.h>
      |          ^~~~~~~~~~~~~~~~~~~
```

ahsan-ca replied:
This seems to be causing this error too:
```
File "/src/AMDMIGraphX/tools/accuracy/accuracy_checker.py", line 26, in
import migraphx
ImportError: /src/AMDMIGraphX/build/lib/libmigraphx_gpu.so.2011000: undefined symbol: hipblasSetStream
```

ahsan-ca replied:
Ahsan to put a comment to add that we are doing this to workaround the broken hipblaslt package.
To Do:
1. Verify with hipBLASLt team if this is not fixed and open an issue to fix the hipblaslt package:
   a. Dev package needs to be fixed to inclue hipblas header files.
   b. Automatically add hipblas in cmake.

ahsan-ca replied:
Added comment to explain its a workaround.

ahsan-ca replied:
Chris checked and the answer to the broken issue is:
```
In 6.2, hipBLAS is a dependency of hipBLASLt, so you should be able to use them together.
In 6.3, the dependencies got change around a bit and you will need to also install a package
called hipBLAS-common.
When you are adding your code now, please don’t assume that just because you have
hipBLASlt installed that means you will have hipBLAS, since this assumption will go away in 6.3.
If you need hipBLAS, install it explicitly and #include it explicitly.
```
This explains how the code in develop that we were looking at would work. They have a new package hipBLAS-common.
```
#ifndef LEGACY_HIPBLAS_DIRECT
#include <hipblas-common/hipblas-common.h>
#else
#include <hipblas/hipblas.h>
#endif
```
@pfultz2 In light of above, do we want to keep hipblas for now since CI is at 6.1. Once we update the CI to 6.3, we can remove hipblas requirement and add `hipBLAS-common` requirement. Does this seem a reasonable approach for now?

ahsan-ca replied:
I missed the step of updating CI to 6.2 before we update it to 6.3 :). I imagine that 6.2 should work fine though (with current workaround), since the changes for hipblaslt side went in, in 6.3.

pfultz2 replied:
Yea we need a comment for the workaround since there package is broken in 6.1(and possibly 6.2 as well).

If hipBLAS-common package is needed in 6.3, then they need to make sure its included by default. We are not using hipblas at all.

pfultz2 commented on tools/docker/sles.docker:
```diff
@@ -23,7 +23,12 @@ RUN zypper --gpg-auto-import-keys install -y \
+    perl-File-BaseDir \
+    libgfortran5 \
+    hipblaslt-devel \
+    hipblas-devel
```
These should also be added to the install_preqs.

ahsan-ca replied:
Done.

pfultz2 commented on tools/install_prereqs.sh:
```diff
@@ -53,6 +53,7 @@ else
     python3-pip \
     python3-venv \
     rocblas-dev \
+    hipblaslt \
```
dont we need to install the dev package?

ahsan-ca replied:
Updated.

ahsan-ca commented on Dockerfile:
```diff
@@ -53,8 +53,10 @@ RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --allow-
+    libgfortran5 \
     hipcub  \
     hipblas  \
+    hipblaslt \
```
Remove this and libgfortran5.

ahsan-ca replied:
Done.

pfultz2 commented on src/targets/gpu/hipblaslt.cpp:
```diff
@@ -0,0 +1,75 @@
+    return hipblaslt_handle_ptr{handle};
+}
+
+hipblaslt_handle_ptr create_hipblaslt_handle_ptr(hipStream_t s)
```
This function can be removed, we already pass the stream when we call hipblasLtMatmul.

ahsan-ca replied:
Sounds good. I will do that.

ahsan-ca replied:
I have removed the function.

bpickrel commented on test/gpu/gemm_tune.cpp:
```diff
@@ -130,7 +131,7 @@ TEST_CASE(gemm_tune_strided)
         }
     }
-#ifdef MIGRAPHX_USE_ROCBLAS_TUNING_API
+#if defined(MIGRAPHX_USE_ROCBLAS_TUNING_API) or defined(MIGRAPHX_USE_HIPBLASLT)
```
After using `target_compile_definitions` to set `MIGRAPHX_USE_HIPBLASLT` in the makefile, it appears to me that it will always be defined and set to either 0 or 1 at this point, which would make this `#if` always true.  Can you confirm that the code really doesn't get compiled when we don't want it to?

ahsan-ca replied:
Thanks for catching this. I will update this.

ahsan-ca replied:
I have updated it to check for the value instead of definition.

bpickrel replied:
I wish we had a convention for preprocessor definitions that would make this mistake less likely, but the fix looks good.

ahsan-ca commented:
The Windows build is passing now.
The PR is good for review, all comments, so far, have been addressed.

pfultz2 commented on src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:
```diff
@@ -0,0 +1,109 @@
+        }
+        return status;
+    });
+}
```
These functions should be in the migraphx::gpu namespace.

ahsan-ca replied:
Done.

pfultz2 commented on src/targets/gpu/include/migraphx/gpu/hip_gemm_impl.hpp:
```diff
@@ -0,0 +1,75 @@
+MIGRAPHX_DECLARE_ENV_VAR(MIGRAPHX_ENABLE_HIP_GEMM_TUNING);
+
+using milliseconds = std::chrono::duration<double, std::milli>;
+using microseconds = std::chrono::duration<double, std::micro>;
```
These shoudl be in the migraphx::gpu namespace as well.

pfultz2 commented on src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:
```diff
@@ -0,0 +1,109 @@
+// provides an API for doing this in hipBLASLt.
+
+// Convert hipblas_status to string
+inline const char* hipblasStatusToString(hipblasStatus_t status)
```
Function names should use snake_case. I dont know why tidy isn't complaining.
